mcos/optimizer.py

```diff
@@ -253,30 +253,40 @@ def _cluster_var(self, cov: np.ndarray) -> np.ndarray:
         w_ = self._inverse_variance_weights(cov).reshape(-1, 1)
         return np.dot(np.dot(w_.T, cov), w_)[0, 0]
 
-    def _hrp_weights(self, cov: np.ndarray, sorted_indices: List) -> np.ndarray:
-        """
-        Gets position weights using hierarchical risk parity
-        :param cov: covariance matrix
-        :param sorted_indices: clustering scheme
-        :return: array of position weights
+    def _hrp_weights(self,cov: np.ndarray, sorted_indices: List) -> np.ndarray:
         """
+      Gets position weights using hierarchical risk parity
+      :param cov: covariance matrix
+      :param sorted_indices: clustering scheme
+      :return: array of position weights, sorted to match the original order passed
+      """
+
         if len(sorted_indices) == 0:
             raise ValueError('sorted_indices is empty')
 
         if len(sorted_indices) == 1:
             return np.array([1.])
 
-        split_indices = np.array_split(np.array(sorted_indices), 2)
+        weights = pd.Series(1, index=sorted_indices)
+        clusters = [sorted_indices]
+
+        while len(clusters) > 0:
+            clusters = [cluster[start:end]
+                                for cluster in clusters
+                                for start, end in ((0, len(cluster) // 2), (len(cluster) // 2, len(cluster)))
+                                if len(cluster) > 1]
 
-        left_var = self._cluster_var(cov[:, split_indices[0]][split_indices[0]])
-        right_var = self._cluster_var(cov[:, split_indices[1]][split_indices[1]])
+            for x in range(0, len(clusters), 2):
```

Comment:
Again please explain why this produces different results from the previous methodology